app/controllers/concerns/foreman/controller/smart_proxy_auth.rb

```diff
@@ -49,11 +50,37 @@ def require_smart_proxy_or_login(features = nil)
   def auth_smart_proxy(proxies = SmartProxy.all, require_cert = true)
     request_hosts = nil
     if request.ssl?
-      dn = request.env[Setting[:ssl_client_dn_env]]
-      if dn && dn =~ /CN=([^\s\/,]+)/i
+      # If we have the client certficate in the request environment we can extract the dn and sans from there
+      # and use the dn in the environment if no certificate is set
+      # SAN validation requires "SSLOptions +ExportCertData" in Apache httpd
+      if request.env.has_key?(Setting[:ssl_client_cert_env]) and !request.env[Setting[:ssl_client_cert_env]].empty?
+        logger.debug "Examining client certificate to extract dn and sans"
+        dn = nil
+        cert_raw = request.env[Setting[:ssl_client_cert_env]]
+        certificate = OpenSSL::X509::Certificate.new(cert_raw)
+        subject_alt_name_extension = certificate.extensions.find {|e| e.oid == "subjectAltName"}
+        if subject_alt_name_extension
```

Comment:
does it make sense to extract to smaller methods? i think its easier to follow the logic this way?